clippy_utils/src/lib.rs

```diff
@@ -621,6 +621,20 @@ fn projection_stack<'a, 'hir>(mut e: &'a Expr<'hir>) -> (Vec<&'a Expr<'hir>>, &'
     (result, root)
 }
 
+/// Checks if the expression is borrowed via an adjustment.
+pub fn expr_borrow_adjustment(cx: &LateContext<'_>, e: &Expr<'_>) -> Option<Mutability> {
+    for a in cx.typeck_results().expr_adjustments(e) {
+        match a.kind {
+            Adjust::Deref(Some(d)) => return Some(d.mutbl),
+            Adjust::Deref(..) => (),
+            Adjust::Borrow(AutoBorrow::Ref(_, m)) => return Some(m.into()),
+            Adjust::Borrow(AutoBorrow::RawPtr(m)) => return Some(m),
+            _ => return None,
```

Comment:
Are the borrow branches necessary? I would think the overloaded deref is what matters and there's an early return there anyways.

Reply:
To fix the problem with `manual_swap`, no. The function name would need to be changed.

Reply:
Done. Also now handles `Box`'s deref.

Reply:
So the borrow branches are only needed for box? What is meant by "custom deref adjustment"? Why not just have `expr_has_overloaded_deref(cx, expr) -> bool`? Anything more specific than that seems pretty niche for clippy_utils.

Reply:
The borrow branches were originally to match the function name, which would have handled just a borrow being added as well (e.g. `x.foo()` can add a borrow to `x`). That isn't needed for `manual_swap` as it will never look at just the receiver of a method call.

`Box` dereferencing is handled the same as a regular reference (Adjust::Deref(None)) rather than through a custom deref, but it still requires borrowing the `Box`, so there's special handling to detect that case. I'm pretty sure the plan is to remove all the special handling for `Box` from rustc at some point, so this can hopefully be removed eventually.

Naming is hard here as it only handles implicit deref's inserted by the compiler, so any name that doesn't mention that is going to be ambiguous. Since they're already called adjustments by rustc there isn't really a reason to call it anything else. I guess `overloaded_deref_adjustment` or `overloaded_implicit_deref` would work.

Reply:
The following compiles, so I think box deref is okay?

```rust
let mut x = Box::new((String::new(), String::new()));
std::mem::swap(&mut x.0, &mut x.1);
```

Reply:
So it does. I remember running into a case where it didn't match regular deref behaviour. I'll remove it and deal with it later if need be.